### ISSUE TITLE
Add `FixedSizeBinary` with support up to at least 32 bytes.

### ISSUE THREAD

**Issue by JayWhite2357** [spaceandtimefdn/sxt-proof-of-sql#229]
# Background and Motivation

Many data are binary blobs. In particular, the EVM address and hash types are 20 bytes and 32 bytes respectively. We should support these natively within Proof of SQL. See https://docs.rs/arrow/latest/arrow/datatypes/enum.DataType.html#variant.FixedSizeBinary.

# Changes Requested

Add full support for a `FixedSizeBinary` type. In order to do this there are several considerations.
* Since `Scalar` types support naturally embedding 31 bytes but not 32 bytes, there should be a divergence in how <=31 bytes are handled and 32 bytes are handled.
    * 31 bytes and under should be naturally embedded into the `Scalar` type. This means simply relying on the conversion from `[u64; 4]` to the `Scalar`.
    * 32 bytes should be handled by first hashing the data and then converting it to a `Scalar`. This is identical to how `String`s are handled. See https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/48b2f0bc8b92dbc3ef9deecfdfc04dfbc2b4460c/crates/proof-of-sql/src/base/scalar/mont_scalar_from.rs#L16.
* `ColumnType` should probably have a new variant `FixedSizeBinary(i32)` added. It is not as clear what the variant for `Column` and `OwnedColumn` should be.
    * My initial suggestion would be `FixedSizeBinary(i32, &'a [u8])` and `FixedSizeBinary(i32, Vec<u8>)` respectively, but that may not make sense. The reason for this suggestion is because something like `Vec<Vec<u8>>` has a lot of overhead, and `Vec<[u8]>` is not allowed by Rust.
    * Alternatively, it may make sense to have two variants: one for the <=31 bytes case, and one for the 32 bytes case. In this case, it may make sense to have the latter simply be `Binary` instead of `FixedSizeBinary`.
* #228 is partially motivated by this issue, and may make this issue a bit more ergonomic. However, working on them in parallel will likely lead to some branch conflict, so pay attention to that issue.
* The only expression that needs to be supported is equality (`=`) and inequality (`!=`). This is because there is no way to add, subtract, etc. between binary data. Eventually, we may wish to add bitwise operations, but those are beyond the scope of this issue.
* It is not clear what the best way to handle binary literals is. We strive to adhere to PostgreSQL syntax, but this data type is unsupported in PostgreSQL. See https://www.postgresql.org/docs/current/datatype-binary.html. There is some flexibility here, but there should be agreement on this before this particular implementation is done.

**Comment by tareknaser** [comment]
Hello @JayWhite2357 

I’d be happy to take on this issue. I think it could be broken down into multiple PRs for smaller, easier-to-test patches. Let me know if that works for you, or if you’d prefer to handle it all in one go.

  

I’ve just joined the Discord and would be happy to discuss it further.

**Comment by JayWhite2357** [comment]
@tareknaser I agree that this makes sense to be broken into smaller PRs. We much prefer smaller PRs over large PRs.

**Comment by JayWhite2357** [comment]
It looks like there are multiple people working on this issue. It may be wise for you to coordinate via Discord.
In the case where there are multiple PRs **that get merged** from different contributors completing the issue, I will split the bounty according to my best judgement. (Unless the contributors have already agreed on a split.)
Alternatively, someone can break this into smaller issues, and I can assign smaller bounties to each.

**Comment by tareknaser** [comment]
It’s a bit tricky to break the issue into clear pieces right now. I think it’ll become clearer once we begin the actual implementation. I can start with a small PR to implement `FixedSizeBinary` to set the stage for follow-up PRs. During the PR review, we can probably get a better sense of how to proceed.

Would that be okay with you, @Ashu999 ?

**Comment by Ashu999** [comment]
sure, go ahead @tareknaser

**Comment by JayWhite2357** [comment]
Testing out two other bounty program options here:
/boss $1000 verify

https://gitpay.me/#/task/974

**Comment by pnogemane** [comment]
Hi @JayWhite2357  -  Is this issue still available to pick up?

**Comment by Dustin-Ray** [comment]
I believe this issue is targeted for internal resolution, thanks for the interest everyone

**Comment by tareknaser** [comment]
> I believe this issue is targeted for internal resolution, thanks for the interest everyone

That’s fine, but just to clarify, it was listed as a bounty, and I’ve already put a lot of time into it. I worked on this PR: [https://github.com/spaceandtimelabs/sxt-proof-of-sql/pull/240](https://github.com/spaceandtimelabs/sxt-proof-of-sql/pull/240), got a review, and even had to open another PR in `blitzar-rs` ([https://github.com/spaceandtimelabs/blitzar-rs/pull/44](https://github.com/spaceandtimelabs/blitzar-rs/pull/44)), which got merged. Since then, I’ve updated the code, rebased, and have been asking for another review round with no response.

I think it would be fair to get a tip for the work done so far. [Algora supports tipping](https://docs.algora.io/commands), so I hope you’ll consider it.

That said, it’s totally your call if you decide not to. No hard feelings either way

**Comment by Dustin-Ray** [comment]
> That said, it’s totally your call if you decide not to. No hard feelings either way

I didnt realize there was already work done on this, it was suggested as something to self assign today. Ill check in with the team and we will get back to you ASAP. The branch is out of date at the moment and I can see the failing test. You dont have to but if you could bring it up to speed that might be nice to get things moving along. 

I agree with you, if you've already put effort into this, there's no point in duplicating the work

**Comment by JayWhite2357** [comment]
@Dustin-Ray and @tareknaser Can the two of you coordinate this. When it gets done, I'll make sure to tip/reward @tareknaser.
Apologies for the slow responses. I didn't realize this was still active.

**Comment by Dustin-Ray** [comment]
> [@Dustin-Ray](https://github.com/Dustin-Ray) and [@tareknaser](https://github.com/tareknaser) Can the two of you coordinate this.

Absolutely, sounds like a plan

**Comment by tareknaser** [comment]
> Can the two of you coordinate this.

Sure

I just rebased the branch on top of main. A lot of code has been added since October when I last worked on this, so I’ve added some `todo!()`s in those areas to get the code to compile (so we can see failing tests). My goal is for this to be a starting point for discussion so we can decide how to move forward


> Please add me on discord

Done

**Comment by yihangwu539-png** [comment]
I'll work on this bounty (Add FixedSizeBinary support). Claiming it now. 👨‍💻

**Comment by yihangwu539-png** [comment]
I'll work on this bounty. Claiming it. 👨‍💻

**Comment by yihangwu539-png** [comment]
Working on this! Draft PR: https://github.com/yihangwu539-png/sxt-proof-of-sql/pull/1

Core type system changes are done (ColumnType, OwnedColumn, Column, LiteralValue variants + Arrow conversions). Still fixing some match branch exhaustiveness errors across ~11 more files. Will update once compilation passes.

**Comment by Ghostofcaldera** [comment]
Hi! I've analyzed this issue and I'd like to work on it.

**Root cause identified:** Implement `FixedSizeBinary` by extending `Scalar` to handle <=31 bytes directly and hash 32-byte inputs. Add a new `ColumnType` variant and update `Column`/`OwnedColumn` to store fixed-size binary dat

I'm planning to submit a PR within the next few hours. Does this approach sound right, or are there constraints I should know about?

**Comment by Ghostofcaldera** [comment]
Looked into this issue. Implement `FixedSizeBinary` by extending `ColumnType` with a new variant and handling <=31 bytes via direct embedding in `Scalar` and 32 bytes via has

Happy to put together a PR unless someone's already working on it.

**Comment by Ghostofcaldera** [comment]
Looked into this issue. Implement a new `FixedSizeBinary` type in the `ColumnType` enum with variants for <=31 bytes (embedded in `Scalar`) and 32 bytes (hashed to `Scalar`).

Happy to put together a PR unless someone's already working on it.

**Comment by Ghostofcaldera** [comment]
Been debugging this. Implement a new `FixedSizeBinary` type in the `ColumnType` enum with variants for <=31 bytes (embedded in `Scalar`) and 32 bytes (hashed to `Scalar`).

Got a fix in mind unless there's a preferred approach.

**Comment by JayWhite2357** [comment]
Closing all issues due to suspected low-effort AI slop/spam.
